R4R: Store Refactor 1 (#2985)
cosmos/cosmos-sdk: store/README.md

## CacheMulti

`cachemulti.Store` is a wrapper `MultiStore` which provides buffered writing / cached reading functionalities over the underlying `MutliStore`

```go
type Store struct {
    db types.CacheKVStore
    stores map[types.StoreKey] types.CacheWrap
}
```

`cachemulti.Store` cache wraps all substores in its constructor and hold them in `Store.stores`. `Store.GetKVStore()` returns the store from `Store.stores`, and `Store.Write()` recursively calls `CacheWrap.Write()` on the substores.

## DBAdapter

`dbadapter.Store` is a adapter for `dbm.DB` making it fulfilling the `KVStore` interface.

```go
type Store struct {
    dbm.DB
}
```

`dbadapter.Store` embeds `dbm.DB`, so most of the `KVStore` interface functions are implemented. The other functions(mostly miscellaneous) are manually implemented.

## IAVL

`iavl.Store` is a base-layer self-balancing merkle tree. It is guaranteed that 

> 1. get/set operation is done within `O(log n)`

Reviewer: ```suggestion
1. Get & set operations are `O(log n)`, where `n` is the number of elements in the tree
```

> 2. iteration efficiently returns the sorted elements within the range

Reviewer: ```suggestion
2. Iteration efficiently returns the sorted elements within the range
```

> 3. each version is immutable, can be retrieved even after a commit

Reviewer: ```suggestion
3. Each tree version is immutable and can be retrieved even after a commit
```

Reviewer: (and this is dependent on pruning settings)

Specification and implementation of IAVL tree can be found in [https://github.com/tendermint/iavl].

## GasKV

`gaskv.Store` is a wrapper `KVStore` which provides gas consuming functionalities over the underlying `KVStore`.

```go
type Store struct {
    gasMeter types.GasMeter
    gasConfig types.GasConfig
    parent types.KVStore
}
```

When each `KVStore` methods are called, `gaskv.Store` automatically consumes appropriate amount of gas depending on the `Store.gasConfig`.

## Prefix

`prefix.Store` is a wrapper `KVStore` which provides automatic key-prefixing functionalities over the underlying `KVStore`.

```go
type Store struct {
    parent types.KVStore
    prefix []byte
}
```

When `Store.{Get, Set}()` is called, the store forwards the call to its parent, with the key prefixed with the `Store.prefix`.

> When `Store.Iterator()` is called, it does not simply prefix the `Store.prefix`, since it does not work as intended. 

Reviewer: What does it do instead? Let's explain.

## RootMulti

> `rootmulti.Store` is a base-layer `MultiStore` where multiple `KVStore` can be mounted on it and retrieved via object-capability keys

Reviewer: Let's mention that the keys are memory addresses.
